### PR TITLE
Correct null dereference exception on unusual import URIs

### LINKED ISSUE #2426: NPE: importedExportedLibraries was called on null
Affected version: `0.36.2`
Original report: https://github.com/dart-lang/pub-dev/issues/4238

To reproduce:
- Download `flutter_instagram_stories` from [here](https://storage.googleapis.com/pub-packages/packages/flutter_instagram_stories-0.1.5%2B4.tar.gz).
- Run `dartdoc` without any extra parameters (but with Flutter in path or `FLUTTER_ROOT` set).

Also affected packages: `fid_widgets`, `typeahead`.

````
NoSuchMethodError: The getter 'importedExportedLibraries' was called on null.
Receiver: null
Tried calling: importedExportedLibraries
#0      Object.noSuchMethod (dart:core-patch/object_patch.dart:51:5)
#1      Library.importedExportedLibraries (package:dartdoc/src/model/library.dart:272:47)
#2      Library.packageImportedExportedLibraries.<anonymous closure> (package:dartdoc/src/model/library.dart:253:52)
#3      Iterable.forEach (dart:core/iterable.dart:283:30)
#4      Library.packageImportedExportedLibraries (package:dartdoc/src/model/library.dart:252:12)
#5      _MarkdownCommentReference._reducePreferLibrariesInLocalImportExportGraph.<anonymous closure> (package:dartdoc/src/markdown_processor.dart:531:24)
#6      SetMixin.any (dart:collection/set.dart:190:15)
#7      _MarkdownCommentReference._reducePreferLibrariesInLocalImportExportGraph (package:dartdoc/src/markdown_processor.dart:530:17)
#8      _MarkdownCommentReference.computeReferredElement (package:dartdoc/src/markdown_processor.dart:471:21)
#9      _getMatchingLinkElement (package:dartdoc/src/markdown_processor.dart:240:12)
#10     _makeLinkNode (package:dartdoc/src/markdown_processor.dart:855:16)
#11     new MarkdownDocument.withElementLinkResolver.linkResolver (package:dartdoc/src/markdown_processor.dart:950:14)
````

### PR DESCRIPTION
**Pull request by jcollins-g**
Fixes #2426.

**BREAKING CHANGE** for PackageGraph.allLibraries (change of key type).

Import URIs that resolve via unusual paths, but otherwise map to the "same" library, result in a null pointer exception when traversing the import/export graph.  Fix this by forcing resolution to `LibraryElement`s generated for the normalized URIs.

### CODE REVIEW

**Comment by CaptainIRS** [comment]
I think it would be useful to emit a `no-defining-library-found` warning when an unusual URI is encountered. https://github.com/dart-lang/dartdoc/blob/6d3c36094d1efba35ce99da8c3d0feaedebdc358/lib/src/warnings.dart#L152-L157

File: lib/src/model/package_graph.dart [hunk 1]
```diff
@@ -869,32 +876,29 @@ class PackageGraph {
...
-    }
+    final libraryElement = resolvedLibrary.library;
     // can be null if e is for dynamic
-    if (elementLibrary == null) {
+    if (libraryElement == null) {
       return null;
     }
-    var foundLibrary = Library.fromLibraryResult(
+    var foundLibrary;
+    foundLibrary = findButDoNotCreateLibraryFor(libraryElement);
```
Comment (srawlins) [inline lib/src/model/package_graph.dart:892]: Why declare this on line 891 and then initialize it on 892?

**Review by srawlins — APPROVED** [review]

File: lib/src/model/package_graph.dart (hunk 1 above)
Comment (jcollins-g) [reply lib/src/model/package_graph.dart]: No good reason, I just rewrote these sections of code too many times working on this.  Fixed.

**Comment by jcollins-g** [comment]
> I think it would be useful to emit a warning like the `no-defining-library-found` warning when an unusual URI is encountered.
> 
> https://github.com/dart-lang/dartdoc/blob/6d3c36094d1efba35ce99da8c3d0feaedebdc358/lib/src/warnings.dart#L152-L157

I'm not entirely opposed to the idea, but I wonder if that might better belong in the linter.  Filed as #2431.

**Pull request merged**